src/aproxy.py

```diff
@@ -0,0 +1,294 @@
+
+        target_proxy = f"{self.config.proxy_address}:{self.config.proxy_port}"
+        if not self._is_proxy_reachable(self.config.proxy_address, self.config.proxy_port):
+            logger.error("Proxy is not reachable at %s", target_proxy)
+            raise ConnectionError(f"Proxy is not reachable at {target_proxy}")
+
+        logger.info("Configuring snap: proxy=%s", target_proxy)
+        snap_cache = snap.SnapCache()
+        snap_cache[APROXY_SNAP_NAME].set({"proxy": target_proxy})
+
+    def _is_proxy_reachable(self, host: str, port: int = DEFAULT_PROXY_PORT) -> bool:
+        """Check if the target proxy is reachable on the specified port.
+
+        Args:
+            host: The target proxy hostname or IP address.
+            port: The port number to check (default is 3128).
+        """
+        try:
+            with socket.create_connection((host, port), timeout=5):
+                return True
+        except (socket.timeout, ConnectionRefusedError, OSError) as e:
+            logger.error("Proxy %s:%s is not reachable: %s", host, port, e)
+            return False
+
+    # ---------------- nftables ----------------
+
+    def _get_primary_ip(self) -> str:
+        """Detect machine's primary IP for DNAT target."""
+        try:
+            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
+                s.connect(("8.8.8.8", 80))
+                return s.getsockname()[0]
+        except (socket.timeout, ConnectionRefusedError, OSError) as e:
+            logger.warning("Falling back to 127.0.0.1 for DNAT target: %s", e)
+            return "127.0.0.1"
+
+    def _resolve_hostname_to_ips(self, hostname: str) -> list[str]:
+        """Resolve a hostname to its corresponding IP addresses.
+
+        Args:
+            hostname: The hostname to resolve.
+
+        Returns:
+            A list of resolved IP addresses.
+        """
+        try:
+            return list(
+                {
+                    str(info[4][0])
+                    for info in socket.getaddrinfo(hostname, None, proto=socket.IPPROTO_TCP)
+                }
+            )
+        except socket.gaierror as e:
+            logger.error("Failed to resolve hostname %s: %s", hostname, e)
+            return []
+
+    def _get_no_proxy_ips(self) -> list[str]:
+        """Convert no-proxy entries to a list of IP addresses.
+
+        Returns:
+            A list of resolved IP addresses.
+        """
+        ip_list: list[str] = []
+        for entry in map(str.strip, self.config.no_proxy.split(",")):
+            if not entry:
+                continue
+            try:
+                # Check if entry is a valid IP or CIDR
+                ipaddress.ip_network(entry, strict=False)
+                ip_list.append(entry)
+            except ValueError:
+                # Not an IP, attempt to resolve as hostname
+                resolved_ips = self._resolve_hostname_to_ips(entry)
+                ip_list.extend(resolved_ips)
```

Comment:
No Proxy support defining domains directly, you don't need to resolve the IP addresses (it also support the following structure (`.canonical.com` which means any subdomain of canonical.com)